src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -48,22 +71,51 @@ namespace NovelRT::Graphics
     GraphicsResourceManager& GraphicsResourceManager::operator=(GraphicsResourceManager&& other) noexcept
     {
         _graphicsDevice = std::move(other._graphicsDevice);
-        _stagingBuffer = Utilities::Lazy<std::shared_ptr<GraphicsBuffer>>([&]() { return CreateStagingBuffer(); });
-
-        if (other._stagingBuffer.isCreated())
-        {
-            _stagingBuffer.reset(other._stagingBuffer.getActual());
-            other._stagingBuffer.reset();
-        }
-
+        _stagingBuffers = std::move(other._stagingBuffers);
+        _defaultBuffers = std::move(other._defaultBuffers);
         _vertexBuffers = std::move(other._vertexBuffers);
+        _indexBuffers = std::move(other._indexBuffers);
+        _constantBuffers = std::move(other._constantBuffers);
         _textures = std::move(other._textures);
         _stagingBufferSize = other._stagingBufferSize;
         other._stagingBufferSize = 0;
+        _contextIndex = other._contextIndex;
+        other._contextIndex = 0;
 
         return *this;
     }
 
+    void GraphicsResourceManager::PrepForFrameWithContextIndex(size_t newContextIndex)
+    {
+        size_t finalContextCount = _contextIndex < newContextIndex ? newContextIndex : _contextIndex;
+
+        size_t currentStagingBufferContextSize = _stagingBuffers.size();
+        while (currentStagingBufferContextSize <= finalContextCount)
+        {
+            _stagingBuffers.emplace_back(std::vector<std::shared_ptr<GraphicsBuffer>>{});
+            ++currentStagingBufferContextSize;
+        }
+
+        if (_stagingBuffers[_contextIndex].size() == 0)
+        {
+            _stagingBufferSize = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
+        }
+        else
+        {
+            size_t totalSize = 0;
+
+            for (auto&& buffer : _stagingBuffers[_contextIndex])
+            {
+                totalSize += buffer->GetSize();
+            }
+
+            totalSize /= _stagingBuffers.size();
+            _stagingBufferSize = totalSize;
+        }
+
+        _stagingBuffers[newContextIndex].clear();
```

Comment:
Is this supposed to be `finalContextIndex`?

Reply:
No because this is for the new context that is getting registered.

What is finalContextIndex anyway? I don't see it in the code.

Reply:
I misread and got confused between finalContextCount and newContextIndex. My bad :sweat_smile: